howto/finetune_adapter_v2.md

```diff
@@ -0,0 +1,105 @@
+# Finetuning with Adapter v2
+
+[LLaMA-Adapter v2](https://arxiv.org/abs/2304.15010) is a form of prefix-tuning that prepends a learnable adaption-prompt to the inputs of the attention blocks in LLaMA. In total, there are only 4M parameters to update during finetuning, which significantly reduces the memory footprint and speeds up training.
+
+We are able to demonstrate instruction-finetuning Lit-LLaMA 7B on the [Alpaca](https://github.com/tatsu-lab/stanford_alpaca) dataset on a **single GTX 3090 (24GB) GPU**. If using 8 GPUs, finetuning can be completed in under 1 hour.
+
+If you are new to LLaMA-Adapter and are interested to learn more about how it works before proceeding with the finetuning guide below, you might find our article [Understanding Parameter-Efficient Finetuning of Large Language Models: From Prefix Tuning to LLaMA-Adapters](https://lightning.ai/pages/community/article/understanding-llama-adapters/) helpful.
+
+LLaMA-Adapter v2 extends the original LLaMA-Adapter idea by adding trainable bias and scale parameters to each linear layer in the transformer. Furthermore, LLaMA-Adapter v2 makes the normalization layers trainable. Where the 7B LLaMA model has 1.2M trainable parameters with LLaMA v1, LLaMA-Adapter v2 adds an additional 2.8 M trainable parameter for the bias and scale parameters and ~300k trainable parameters for the normalization layers. So, adapter v2 has ~4.3 M trainable parameters in total.
```

Comment:
When would the user choose v1 over v2?

If the answer is never, should we replace the v1 implementation?
Otherwise, this should be explained in this howto

Reply:
I'd say that it makes sense to have both methods because Adapter v2 is more heavy-weight than v1. E.g. where v1 has 1.2 M trainable parameters, v2 has 4.3 M trainable parameters, which is almost 4x as much.

The parameter number is mentioned in the v2 how-to but not in the context of using v1 vs v2. I am happy to add that. 

Do you by chance have any additional ideas that come to mind regarding which version to use?

Reply:
One thing to highlight in the how-to is that the number of trainable parameters changes to 4.4, but inference cost doesn't change that much since using bias & scale parameters in the forward method are cheap, and the RMSNorm parameters (300k) were already there anyways.

Reply:
Good points, we should add those.

I haven't dove deep into the paper, but it describes v2 as an improved method for both multi-model data and text-only data:

> The newly designed framework also
exhibits stronger language-only instruction-following capabilities and even excels in chat interactions